### PR TITLE
[WIP] Wrap C++17 std::filesystem::path to solve encoding issue on Windows.

### LINKED ISSUE #13103: Invalid wallet path with Chinese characters in windows
Reported in https://github.com/bitcoin/bitcoin/pull/13100#issuecomment-384836977 by @ken2812221.

### PR DESCRIPTION
**Pull request by ken2812221**
Fix #13103
We can call `.u8string()` everywhere and call `.string()` when we need to use native-encoded path.

### CODE REVIEW

**Comment by laanwj** [comment]
Neat.
Concept ACK!

**Comment by fanquake** [comment]
Travis macOS build failed with:
```
In file included from util.cpp:6:
In file included from ./util.h:18:
In file included from ./fs.h:11:
In file included from /home/travis/build/bitcoin/bitcoin/depends/x86_64-apple-darwin11/share/../include/boost/filesystem.hpp:16:
/home/travis/build/bitcoin/bitcoin/depends/x86_64-apple-darwin11/share/../include/boost/filesystem/path.hpp:141:33: error: no type named 'type' in 'boost::enable_if<boost::filesystem::path_traits::is_pathable<Path>, void>'; 'enable_if' cannot be used to disable this declaration
      typename boost::enable_if<path_traits::is_pathable<
                                ^~~~~~~~~~~~~~~~~~~~~~~~~
./util.h:315:36: note: in instantiation of member function 'boost::filesystem::path::path' requested here
    using boost::filesystem::path::path;
                                   ^
util.cpp:1098:12: note: while substituting deduced template arguments into function template 'Path' [with Source = Path]
    return Path(Utf8ToNative(source));
           ^
1 error generated.
```

File: src/util.h [hunk 1]
```diff
@@ -309,4 +309,13 @@ std::unique_ptr<T> MakeUnique(Args&&... args)
  */
 int ScheduleBatchPriority(void);
 
+class Path : public boost::filesystem::path
```
Comment (laanwj) [inline src/util.h]: Would be better in `src/fs.h`?

File: src/util.cpp [hunk 2]
```diff
@@ -1020,3 +1022,45 @@ int ScheduleBatchPriority(void)
     return 1;
 #endif
 }
+
+static std::string NativeToUtf8(const std::string& source)
```
Comment (laanwj) [inline src/util.cpp]: string encoding should probably be in `utilstrencodings.[ch]`

**Comment by laanwj** [comment]
> Fix #13103

How does this fix that issue? 
This adds a class and some functions, but never uses them, or am I missing something?

> We can call .u8string() everywhere and call .string() when we need to use native-encoded path.

Concept ACK, though I'm not entirely convinced on wrapping versus using functions where needed.

**Comment by ken2812221** [comment]
I'm trying to re-do #13107 with pretty code, but it's still WIP. The target is to fix #13103

**Comment by ken2812221** [comment]
Seems not possible to wrap all filesystem classes. Close it and reopen #13107

**Pull request closed without merging**